cpp/include/cuvs/neighbors/ivf_pq_helpers.hpp

```diff
@@ -0,0 +1,274 @@
+ *   ivf_pq::helpers::extend_list_with_codes(
+ *       res, &index, codes.view(), indices.view(), label);
+ * @endcode
+ *
+ * @param[in] res
+ * @param[inout] index
+ * @param[in] new_codes flat PQ codes, one code per byte [n_rows, index.pq_dim()]
+ * @param[in] new_indices source indices [n_rows]
+ * @param[in] label the id of the target list (cluster).
+ */
+void extend_list_with_codes(
+  raft::resources const& res,
+  index<int64_t>* index,
+  raft::device_matrix_view<const uint8_t, uint32_t, raft::row_major> new_codes,
+  raft::device_vector_view<const int64_t, uint32_t, raft::row_major> new_indices,
+  uint32_t label);
+
+/**
+ * @brief Extend one list of the index in-place, by the list label, skipping the classification
+ * step.
+ *
+ *  Usage example:
+ * @code{.cpp}
+ *   // We will extend the fourth cluster
+ *   uint32_t label = 3;
+ *   // We will extend with 4 new vectors
+ *   uint32_t n_vec = 4;
+ *   // Indices of the new vectors
+ *   auto indices = raft::make_device_vector<uint32_t>(res, n_vec);
+ *   ... fill the indices ...
+ *   auto new_vectors = raft::make_device_matrix<float, uint32_t, row_major> new_codes(
+ *       res, n_vec, index.dim());
+ *   ... fill vectors ...
+ *   // extend list with new vectors
+ *   ivf_pq::helpers::extend_list(
+ *       res, &index, new_vectors.view(), indices.view(), label);
+ * @endcode
+ *
+ *
+ * @param[in] res
+ * @param[inout] index
+ * @param[in] new_vectors data to encode [n_rows, index.dim()]
+ * @param[in] new_indices source indices [n_rows]
+ * @param[in] label the id of the target list (cluster).
+ */
+void extend_list(raft::resources const& res,
+                 index<int64_t>* index,
+                 raft::device_matrix_view<const float, uint32_t, raft::row_major> new_vectors,
+                 raft::device_vector_view<const int64_t, uint32_t, raft::row_major> new_indices,
+                 uint32_t label);
+void extend_list(raft::resources const& res,
+                 index<int64_t>* index,
+                 raft::device_matrix_view<const int8_t, uint32_t, raft::row_major> new_vectors,
+                 raft::device_vector_view<const int64_t, uint32_t, raft::row_major> new_indices,
+                 uint32_t label);
+void extend_list(raft::resources const& res,
+                 index<int64_t>* index,
+                 raft::device_matrix_view<const uint8_t, uint32_t, raft::row_major> new_vectors,
+                 raft::device_vector_view<const int64_t, uint32_t, raft::row_major> new_indices,
+                 uint32_t label);
+/**
+ * @brief Remove all data from a single list (cluster) in the index.
+ *
+ * Usage example:
+ * @code{.cpp}
+ *   // We will erase the fourth cluster (label = 3)
+ *   ivf_pq::helpers::erase_list(res, &index, 3);
+ * @endcode
+ *
+ *
+ * @param[in] res
+ * @param[inout] index
+ * @param[in] label the id of the target list (cluster).
+ */
+void erase_list(raft::resources const& res, index<int64_t>* index, uint32_t label);
+
```

Comment:
Looks like we are missing the `reset_index` helper. That helper is needed by FAISS. It can also be moved to the index class instead of being an external helper.

Reply:
We are missing `unpack_contiguous_list_data` and `recompute_internal_state` in the helpers.

Reply:
`extract_centers` is missing.

Reply:
Can you ensure that all the helpers have been migrated? They are essential for FAISS' future dependence on cuVS.

Reply:
cc @lowener- we will need to add these in follow-on PRs. We needed to merge this PR for other reasons.